modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/time/AbstractTimeFunction.java

```diff
@@ -0,0 +1,294 @@
+        if (dateTime != null) {
+            return dateTime;
+        }
+        long timeData = timeStruct.getIntField(0);
+        BStruct zoneData = (BStruct) timeStruct.getRefField(0);
+        ZoneId zoneId;
+        if (zoneData != null) {
+            String zoneIdName = zoneData.getStringField(0);
+            zoneId = ZoneId.of(zoneIdName);
+        } else {
+            zoneId = ZoneId.systemDefault();
+        }
+        dateTime = Instant.ofEpochMilli(timeData).atZone(zoneId);
+        timeStruct.addNativeData(KEY_ZONED_DATETIME, dateTime);
+        return dateTime;
+    }
+
+    private BStruct createCurrentTimeZone(Context context) {
+        //Get zone id
+        ZoneId zoneId = ZoneId.systemDefault();
+        String zoneIdName = zoneId.toString();
+        //Get offset in seconds
+        ZoneOffset o = OffsetDateTime.now().getOffset();
+        int offset = o.getTotalSeconds();
+        return createBStruct(getTimeZoneStructInfo(context), zoneIdName, offset);
+    }
+
+    private BStruct createTimeZone(Context context, String zoneIdValue) {
+        String zoneIdName;
+        try {
+            ZoneId zoneId = ZoneId.of(zoneIdValue);
+            zoneIdName = zoneId.toString();
+            //Get offset in seconds
+            TimeZone tz = TimeZone.getTimeZone(zoneId);
+            int offsetInMills  = tz.getOffset(new Date().getTime());
+            int offset = offsetInMills / 1000;
+            return createBStruct(getTimeZoneStructInfo(context), zoneIdName, offset);
+        } catch (ZoneRulesException e) {
+            throw new BallerinaException("invalid timezone id " + zoneIdValue);
+        }
+
+    }
+
+    private BStruct createTimeZone(Context context, ZoneId zoneId) {
+        //Get zone id
+        String zoneIdName = zoneId.toString();
+        //Get offset in seconds
+        TimeZone tz = TimeZone.getTimeZone(zoneId);
+        int offsetInMills  = tz.getOffset(new Date().getTime());
+        int offset = offsetInMills / 1000;
+        return createBStruct(getTimeZoneStructInfo(context), zoneIdName, offset);
+    }
+
+    private StructInfo getTimeZoneStructInfo(Context context) {
+        StructInfo result = zoneStructInfo;
+        if (result == null) {
+            PackageInfo timePackageInfo = context.getProgramFile().getPackageInfo(TIME_PACKAGE);
+            zoneStructInfo = timePackageInfo.getStructInfo(STRUCT_TYPE_TIMEZONE);
+        }
+        return zoneStructInfo;
+    }
+
+    private StructInfo getTimeStructInfo(Context context) {
+        StructInfo result = timeStructInfo;
+        if (result == null) {
+            PackageInfo timePackageInfo = context.getProgramFile().getPackageInfo(TIME_PACKAGE);
+            timeStructInfo = timePackageInfo.getStructInfo(STRUCT_TYPE_TIME);
+        }
+        return timeStructInfo;
+    }
+
+    private BStruct createBStruct(StructInfo structInfo, Object... values) {
```

Comment:
We can remove this and use method defined in org.ballerinalang.bre.bvm.BLangVMErrors.createBStruct,